### PR TITLE
FT_MOTION : updating last_direction_bits for Core printer solves homing failure

### LINKED ISSUE #26683: [BUG] FT_MOTION Homing and Moving Axes
### Did you test the latest `bugfix-2.1.x` code?

Yes, and the problem still exists.

### Bug Description

FT_MOTION on a CoreXY printer with A and B motors, and 4 Z-Steppers. When FT_MOTION is enabled (time-based motion) the printer will not home. The print head will hit the x-axis limit switch and keep moving in the x-direction while moving to home the y axis. The y-axis does the same thing, crashing into the y-limit switch. The printer will not home the z-axis using the safe z-home coordinates.

Selecting other input shaping modes cause motors to freeze - no motion occurs in X or Y axes.

### Bug Timeline

Dec. 18, 2023 to Jan. 14, 2024

### Expected behavior

x-axis homes, y-axis homes, z-axis homes

### Actual behavior

X and y axis crash into limit switches, z axis prevented from safe homing

### Steps to Reproduce

1. Start Printer
2. Enable FT_Motion Shaping - Fixed-Time Motion Mode set to "ON" from MOTION menu
3. Select "Auto-Home" from motion menu

### Version of Marlin Firmware

Bugfix-2.1.x

### Printer model

Corexy, Voron

### Electronics

BTT Octopus V1.1

### LCD/Controller

FYSETC_MINI_12864_2_1

### Other add-ons

_No response_

### Bed Leveling

UBL Bilinear mesh

### Your Slicer

Cura

### Host Software

None

### Don't forget to include

- [X] A ZIP file containing your `Configuration.h` and `Configuration_adv.h`.

### Additional information & file uploads

[Shewan-1-14-24.zip](https://github.com/MarlinFirmware/Marlin/files/13933032/Shewan-1-14-24.zip)

### LINKED ISSUE #26836: [BUG] Homing Fails on latest bugfix 2.1.x nightly, if FT Motion enabled.
### Did you test the latest `bugfix-2.1.x` code?

Yes, and the problem still exists.

### Bug Description

If Fixed-Time Motion is enabled, printer will move up, then home X, and stop with endstop engaged without backing off, then will not home Y, and not move to the center to probe Z, but probe Z from the location it engaged the X endstop.

### Bug Timeline

Started when first enabled FT Motion

### Expected behavior

Move up, move to X endstop, back off from Endstop, then move to Y endstop, back off from Y, then move Probe to center of bed, and probe bed.

### Actual behavior

Move up, then home X, and stop with endstop engaged without backing off, then will not home Y, and not move to the center to probe Z, but probe Z from the location it engaged the X endstop.

### Steps to Reproduce

M493 S17
G28

### Version of Marlin Firmware

Marlin bugfix-2.1.x (Mar  2 2024 18:32:15) SOURCE_CODE_URL:github.com/MarlinFirmware/Marlin PROTOCOL_VERSION:1.0 MACHINE_TYPE:HEVO 500 KINEMATICS:COREXYCartesian EXTRUDER_COUNT:1 UUID:cede2a2f-41a2-4748-9b12-c55c62f367ff

### Printer model

Hypercube Evolution custom

### Electronics

BTT SKR 1.4 Turbo

### LCD/Controller

BTT TFT70

### Other add-ons

_No response_

### Bed Leveling

ABL Bilinear mesh

### Your Slicer

Prusa Slicer

### Host Software

OctoPrint

### Don't forget to include

- [X] A ZIP file containing your `Configuration.h` and `Configuration_adv.h`.

### Additional information & file uploads

[Configuration.zip](https://github.com/MarlinFirmware/Marlin/files/14471231/Configuration.zip)

### PR DESCRIPTION
**Pull request by narno2202**
### Description
Some users report homing failure with `FT_MOTION` enabled and a CoreXY printer. This PR just disable `FT_MOTION` in G28 when the printer is Core printer. Despite testing no reliable solution has been found, disabling FT_MOTION is just a workaround for now.

### Benefits

Solve #26683

### CODE REVIEW

File: Marlin/src/inc/Conditionals_adv.h [hunk 1]
```diff
@@ -1148,7 +1148,7 @@
   #elif HAS_DRIVER(A4988)
     #define MINIMUM_STEPPER_POST_DIR_DELAY 200
   #elif HAS_TRINAMIC_CONFIG || HAS_TRINAMIC_STANDALONE
-    #define MINIMUM_STEPPER_POST_DIR_DELAY 70
+    #define MINIMUM_STEPPER_POST_DIR_DELAY 100
```
Comment (sargonphin) [inline Marlin/src/inc/Conditionals_adv.h:1188]: Is this change on purpose or is it just a leftover? Same for the change in `stepper.cpp` line 3563

**Comment by narno2202** [comment]
Regarding `MINIMUM_STEPPER_POST_DIR_DELAY,` it's not a leftover (I forgot to mention it in the title), the value is based upon `STEP input minimum low time` from TMC datasheet :  solves the buggy axis motion. The second one is clearly a leftover :)

**Comment by thinkyhead** [comment]
Homing depends on being able to distinguish that a move is occurring in a particular axis, not just with a particular stepper. For Marlin standard Core kinematics motion we set `axis_did_move[X_AXIS_HEAD]`, `axis_did_move[Y_AXIS_HEAD]`, and `axis_did_move[Z_AXIS_HEAD]`.

Meanwhile, `Stepper::ftMotion_stepper` modifies both `axis_did_move` and `last_direction_bits` without consideration for Core kinematics and always overwrites the `hx`, `hy`, and `hz` elements with 0. This effectively breaks the direction checking for endstops, so they will only trigger in the …negative… direction.

The solution is to fill in the `hx`, `hy`, `hz` fields in `axis_did_move` and `last_direction_bits` somewhere, most likely in `ftMotion_stepper`. However, there isn't enough information in `ftMotion.stepperCmdBuff` to determine the direction of the axes based on the directions of the steppers. This is because `axis_did_move` is only set when there is a STEP, and core axes don't necessarily have to step together. Meanwhile, the `last_direction_bits` also don't tell us about the direction of an axis.

It should be possible to add more bit flags to `ftMotion.stepperCmdBuff` to simply indicate that an axis is generally moving in a particular direction in addition to the flags which indicate that an axis should STEP with a DIR. As it currently stands, the usage of `last_direction_bits` by FT Motion is mostly compatible. But `axis_did_move` is constantly toggling as an axis either has a STEP or doesn't have a STEP, which makes it hit-or-miss to discover whether an axis is moving. Under Marlin standard motion `axis_did_move` is set as long as there are steps pending in the current segment.

There might be other features apart from endstop checking that depend on `axis_did_move` and `last_direction_bits`, so for best compatibility we should specifically address Core kinematics in FT Motion and especially make sure that `axis_did_move` is true throughout the whole move.

**Comment by thinkyhead** [comment]
We could possibly set the move and direction flags for `CORE1` and `CORE2` as soon as the new planner block is fetched, assuming that FT Motion immediately starts processing. However, if FT Motion is pre-fetching and populating `ftMotion.stepperCmdBuff` "asynchronously" ahead of completing motion for the previous block then we will need to add more FT Motion command bits, perhaps named `FT_BIT_DIR_C1`/`C2` and `FT_BIT_STEP_C1`/`C2`. I've only added some TODO comments at this point to label the relevant code points.

**Comment by narno2202** [comment]
@thinkyhead , as the directions are valid until the end of  block processing, we could probably set the corresponding fields in` last_direction_bits` when the block is loaded and before his processing begins in `void Stepper::ftMotion_blockQueueUpdate()` (as it was the case in early ft_motion code but without the Core logic).

**Comment by thinkyhead** [comment]
> …we could probably set the corresponding fields in `last_direction_bits` when the block is loaded and before his processing begins…

As long as there's no "pre-fetching" going on that should work. With standard Marlin motion the next block is always fetched after the previous one has completed its STEP pulses. It would be possible for FT Motion to fetch the next block and start adding STEP/DIR events to its own command queue as soon as there's space available in its buffer, even if the block hasn't completed processing yet —which could provide improved continuity— but I wasn't sure if it was already doing that.

**Comment by thinkyhead** [comment]
> …the directions are valid until the end of block processing…

That being the case, do we really need to include DIR information in the FT Motion command buffer, or could the DIR pins also be set just once when the new block is fetched? I gather that including both the DIR and STEP in the same buffer potentially allows for less delay, but most stepper drivers require a delay in-between changing DIR and starting the next STEP anyway.

**Comment by narno2202** [comment]
After some testing, when homing, there are only 3 consecutive loaded blocks per axis : 

1.  Block for the first move to the origin
2.  Block for bumping if enabled
3.  Final block: last move to the origin
So it could be sufficient to set hx, hx and hz in `void Stepper::ftMotion_blockQueueUpdate() `

**Comment by thinkyhead** [comment]
I've done a bit more cleanup on this, adding a shared method `set_axis_moved_for_current_block` used by both standard motion (`block_phase_isr`) and FT motion (`ftMotion_blockQueueUpdate`) when fetching a new block, so any axis that is moving will be flagged at the start and this state will endure through the entire block (rather than toggling based on whether a STEP signal is being generated in the moment). And `ftMotion_stepper` now does its own thing with DIR/STEP without affecting the flags that pertain to the whole block.

**Comment by narno2202** [comment]
@thinkyhead , is `endstops.update()` mandatory outside the homing function?
I made some tests with a flag set in G28 on entry and unset on exit.  hx, hy and hz are set as before in `ftMotion_blockQueueUpdate` only if the flag is true and  `endstops.update()` is only called when the flag is true. Homing is fine and printing too. There is some workload gain.

**Comment by thinkyhead** [comment]
> is `endstops.update()` mandatory outside the homing function?

Endstop states are tracked continually, but only interrupt motion when they are "enabled." They may be enabled during printing for a feature like `SD_ABORT_ON_ENDSTOP_HIT`, or for a custom application that monitors them for some non-standard usage.

**Comment by thisiskeithb** [comment]
> They may be enabled during printing for a feature like `SD_ABORT_ON_ENDSTOP_HIT`

`ENDSTOPS_ALWAYS_ON_DEFAULT` is probably the most common case I see when helping users.

**Comment by thinkyhead** [comment]
> `ENDSTOPS_ALWAYS_ON_DEFAULT` is probably the most common case I see when helping users.

Only 4 of our example configs have this option enabled. It's only recommended for users who really need this behavior for some reason.

**Comment by thisiskeithb** [comment]
> Only 4 of our example configs have this option enabled. It's only recommended for users who really need this behavior for some reason.

Understood, but this is for general support outside of the configs we host.

**Comment by thinkyhead** [comment]
> I made some tests with a flag set in `G28` on entry and unset on exit. `hx`, `hy` and `hz` are set as before in `ftMotion_blockQueueUpdate` only if the flag is true and `endstops.update()` is only called when the flag is true. Homing is fine and printing too. There is some workload gain.

- Endstops are currently enabled during `G28` and then they are returned to their prior state (i.e., they may remain enabled).
- There is a direct call to `endstops.update()` to acquire endstop states within `block_phase_isr` when fetching a new block. This is needed to deal with already-pressed endstops at the start of a new move, which would not trigger an interrupt.
- When endstop interrupts are disabled, the Temperature ISR gets the latest endstop states by calling `endstops.poll()` ~1000 times per second.
- When endstop interrupts are enabled, the Temperature ISR instead just handles debounce of the endstop interrupts.

So there should be no need to call `endstops.update()` in `ftMotion_stepper` (unless there is some need for more accurate triggering than would be achieved within 0.001 seconds). But it seems like a good idea to call `endstops.update()` in `ftMotion_blockQueueUpdate` to match the behavior of `block_phase_isr`.

**Comment by narno2202** [comment]
My vision was too simplistic : home all axes then all the moves are constrained into min and max of each axis, so endstops.update() is not needed until a new G28 is sent.

**Comment by narno2202** [comment]
Currently BLTOUCH_HS mode is broken

**Comment by narno2202** [comment]
`axis_did_move` needs to be updated every time we pop-up a new command from the buffer as we don't know which block it belongs to. Doing this solves BLTOUCH_HS mode failure. `set_axis_moved_for_current_block()` is not needed anymore in `ftMotion_blockQueueUpdate()` as it's broke the whole thing.
I'm not sure that  this conditionnal is mandatory for a standard printer as it's always true : `if (current_block->is_pos_sync()) _set_position(current_block->position);`. I think that the while loop has to be refactored, but I don't feel comfortable with laser feature.

**Comment by thinkyhead** [comment]
> `axis_did_move` needs to be updated every time we pop-up a new command from the buffer as we don't know which block it belongs to.

Previously we updated `axis_did_move` to reflect the current STEP command for each axis:
```cpp
axis_did_move = LOGICAL_AXIS_ARRAY(
  TEST(command, FT_BIT_STEP_E),
  TEST(command, FT_BIT_STEP_X), TEST(command, FT_BIT_STEP_Y), TEST(command, FT_BIT_STEP_Z),
  TEST(command, FT_BIT_STEP_I), TEST(command, FT_BIT_STEP_J), TEST(command, FT_BIT_STEP_K),
  TEST(command, FT_BIT_STEP_U), TEST(command, FT_BIT_STEP_V), TEST(command, FT_BIT_STEP_W)
);
```
It appears that `axis_did_move` will be constantly flipping between 0 and 1 since there will be some blocks with a STEP command for the given stepper and –depending on angle– most commands might not have a STEP command for the axis. So `axis_did_move` cannot itself be reliably used to determine that a particular axis is involved in the current move.

But would this work?
```cpp
const AxisBits didmove = LOGICAL_AXIS_ARRAY(
  TEST(command, FT_BIT_STEP_E),
  TEST(command, FT_BIT_STEP_X), TEST(command, FT_BIT_STEP_Y), TEST(command, FT_BIT_STEP_Z),
  TEST(command, FT_BIT_STEP_I), TEST(command, FT_BIT_STEP_J), TEST(command, FT_BIT_STEP_K),
  TEST(command, FT_BIT_STEP_U), TEST(command, FT_BIT_STEP_V), TEST(command, FT_BIT_STEP_W)
);
axis_did_move |= didmove; // Any axis moved since the last block was fetched?
```

It seems like it should work, since we call `axis_did_move.reset()` before releasing the current block in `ftMotion_blockQueueUpdate`. But that is called from the main loop. So, is it possible for STEP/DIR commands from a previous block to still be running (from the ISR) at the same time as the next block is being fetched (from the main loop)? If there's any overlap then the above code would pick up the tail end of the running block, stepping on the reset from `ftMotion_blockQueueUpdate`. But if there is no overlap, it seemed plausible that calling `set_axis_moved_for_current_block` right when fetching the new block would work.

> I'm not sure that this conditionnal is mandatory for a standard printer as it's always true : `if (current_block->is_pos_sync()) _set_position(current_block->position)`

We have to continue to process position sync blocks (and others). The `is_pos_sync()` method will return `false` for all regular segment blocks, and only returns `true` for blocks that contain updated stepper counts. These blocks are rare, but support for them must be included in the block queue processor to ensure that positions set by high level G-code are applied at the right point in the motion queue.

If blocks are not being fetched just-in-time and there are STEP/DIR commands from a previous block being run at the same time a new block is being fetched, positions will not be in proper sync, so we might have to sacrifice some concurrency or add some new bits to the STEP/DIR command mask to provide for commands like `G92` that will otherwise need to interrupt motion like `M400` does to prevent loss of position sync.

**Comment by narno2202** [comment]
I test the last code, everything seems to be fine. Some changes restore the old code behavior :)
Regarding `if (current_block->is_pos_sync()) _set_position(current_block->position)`, it's always true in the while loop context for a standard printer but not always for a laser machine. The old code call `_set_position(current_block->position) ` for a laser if `is_fan_sync()` or `is_pwr_sync()` is true which is no more the case here unless if `is_pos_sync()` is always true when ` is_fan_sync()` or `is_pwr_sync()` is true. I think that the code could be simplified for a standard printer and tested or reviewed for  a laser machine.

**Comment by thinkyhead** [comment]
> Regarding `if (current_block->is_pos_sync()) _set_position(current_block->position)`, it's always `true` in the `while` loop context for a standard printer but not always for a laser machine.

The position sync flag (`flag.sync_position`) returned by `current_block->is_pos_sync()` should almost never be `true`. The `Planner::_populate_block` method clears all block flags with `block->flag.clear()`. The only time the flag should be `true` is when a block is created by `buffer_sync_block(BLOCK_BIT_SYNC_POSITION)`.

If you're seeing `current_block->is_pos_sync()` returning `true` outside of actual position sync blocks then there must be a bug already existing in the Planner code that we haven't touched.

**Comment by narno2202** [comment]
I'am testing the new code in MarlinCollaborative repository. For now I correct some bugs and compilation went fine, I adapt the FT Motion Menu. Motion seems OK but BLtouch is buggy. Testing underway...

**Comment by ulendoalex** [comment]
For a quick fix, could I suggest using the older implementation of fxdTiCtrl_refreshAxisDidMove() ([here](https://github.com/S2AUlendo/MarlinCollaborative/blob/d8ffae3474588d12064ccbc3c7592ba8ea9d24c6/Marlin/src/module/stepper.cpp#L3550)), with some additional logic for Core XY such as in the [standard motion block phase](https://github.com/S2AUlendo/MarlinCollaborative/blob/d8ffae3474588d12064ccbc3c7592ba8ea9d24c6/Marlin/src/module/stepper.cpp#L2535) ? Or perhaps you already know why this might not work ?

I'm hesitant to get on board with the current direction which is doing the endstops update() every FTM stepper ISR since we usually configure it from 40kHz-120kHz and want to keep it as lightweight as possible.

Another option is calculating axis_did_move at the block load, and propagating the information down with the stepper commands, as @thinkyhead mentioned. Unfortunately this wastes some memory/cycles on data that doesn't change very frequently relative to the stepper frequency.

One more option I can think of is calculating axis_did_move at the block load, and holding any true values until the motion is certainly completed using timers. We could calculate that time using the configuration information (i.e.: batch size, buffer sizes).

Please let me know what you think.

**Comment by thinkyhead** [comment]
@ulendoalex — Thanks for your comments. Here are my initial responses. I'd like to get this and the other FT Motion changes merged soon, but of course not until we've worked out the kinks!

> I suggest using the older implementation of `fxdTiCtrl_refreshAxisDidMove()`… I'm hesitant to get on board with the current direction which is doing the `endstops.update()` every FTM stepper ISR…

I agree we don't need to check the endstops states quite that often, but only when certain axes have stepped. If it is sufficient to poll the states at 1KHz (with slightly less accuracy, but that's what the "slow bump" is for) then we can simply not update the states in the stepper ISR. Endstops will still work, and of course this has no effect on endstops using interrupts, which is actually most 32-bit boards (where most pins are interrupt-capable).

So are you saying that `fxdTiCtrl_refreshAxisDidMove()` should be reverted regardless of the `endstops.update` situation, or only to reduce the number of endstop calls?

> …calculating `axis_did_move` at the block load, and propagating the information down with the stepper commands…

That would seem to be the way to do it, because we need the state to be synchronized with the motion and not with the block processing that happens a bit earlier. It may be that all we really need is a bit that indicates the start of a new block, and then we would just clear `axis_did_move` in response (along with any other "new block" things). Then to set bits in `axis_did_move` we would just OR the STEP bits with it while FT stepping. The bits would remain set throughout the segment, guaranteed to be seen by the temp ISR endstop polling.

> calculating `axis_did_move` at the block load…

We could get the bits we want to add to `axis_did_move` at that time, but we couldn't actually apply them to `axis_did_move` until the block is starting to actually STEP. Therefore we need at minimum a "new block" bit and at most a whole set of "axis moved" bits added to the command stream. I'd prefer to try adding the single "new block" bit and see how that looks, and only add separate "axis moved" bits if the single bit doesn't do the trick.

I'll whip that together shortly and we'll see how it looks….

**Comment by thinkyhead** [comment]
Ok, I've added the new flag to ensure that `axis_did_move` is set in sync with motion and is no longer cleared before the move is done.

Caveat: Adding a single bit should do the trick, but unfortunately makes the most common case (XYZE) go from 8 bit commands to 16-bit commands. So for best optimization we would need to replace `FT_BIT_START` with a separate array of bit flags that tracks along with `stepperCmdBuff_consumeIdx >> 3`.

**Comment by ulendoalex** [comment]
> So are you saying that `fxdTiCtrl_refreshAxisDidMove()` should be reverted regardless of the `endstops.update` situation, or only to reduce the number of endstop calls?

Only to reduce the number of endstop calls. The function at its core was just supposed to do the following:
- In ftMotion_stepper(), any stepping would set axis_did_move true (but lack of a step would NOT clear axis_did_move).
- Every 2.5 ms endstops.update() happens, then axis_did_move gets cleared.
In this way endstops.update() would work off stepping knowledge of the past 2.5 ms. The function would have been a good place to do the kinematics mapping, i.e.: `if CORE_XY && (A_did_step || B_did_step) then axis_did_move.X = true` and `if CORE_XY && (A_did_step != B_did_step) then axis_did_move.Y = true`


The new commit using FT_BIT_START looks like it it would get tripped up with multiple blocks in the planner, but should work well for homing.

What do you think about setting bits in convertToSteps() instead? Set based on the delta variable (applying CORE logic as necessary) and clearing based on time (FTM_STEPPERCMD_BUFF_SIZE/FTM_STEPPER_FS seconds). EDIT: I'll try something like this, but may take a couple days to get a corexy printer setup to test.

**Comment by narno2202** [comment]
Once the printer has been homed, is `endstops.poll() ` in `temperature.isr()` enough to manage ancillary functions related to endstops ? If yes, we have only to set a flag on `homeaxis(const AxisEnum axis)` entry, test it in `ftMotion_stepper()`,  and unset the flag when homeaxis is done. Seems OK on my tests.

**Comment by narno2202** [comment]
What is the status of current bugfix regarding `FT_MOTION` based on users report? :
 - Homing failure for CoreXY printers : solved by setting hx, hy and hz in `ftMotion_blockQueueUpdate()`
 - Inconsistent motion for CoreXY printer with TMC22xx : solved by setting` MINIMUM_STEPPER_POST_DIR_DELAY` to 100ns (TMC datasheet)
 For now, do we really need frequent `endstops.update()` call once the printer has been homed and is` endstops.poll()` enough to manage situation where endstops could be triggered regardless of motion direction?

**Comment by narno2202** [comment]
I am currently running last bugfix with this PR plus all the new Marlincollaborative code (with very minor changes) and an adapted FT_MOTION menu. Motions are fine, sensorless homing is working, BLTouch HS mode too, printing is OK with or without shaping.

**Comment by Mjankor** [comment]
Thanks narno. Was just passing through to see where this PR was up to. Got a CoreXY here that doesn't play nice with current input shaping.

**Comment by ulendoalex** [comment]
@thinkyhead @narno2202 

I have been doing some testing to resolve endstops [homing] issues and am going to propose the following as a solution. This is a timer based method that sets the axis movement end time when the block gets loaded by considering the movement time and time it takes the block to get converted and propagate through FTM's buffers. It requires some changes to endstops.cpp but is otherwise fairly clean.

I also tried a method that watches the FTM stepper(), similar to how refreshAxisDidMove() used to work. However, it did not play nicely with endstop interrupts. In particular, for the case when a block is loaded with motion away from an endstop that is already activated, commands has not yet propagated to the buffer stepper() is using, and the initial endstop check fails as there is no indication the axis is moving. The endstop isn't checked again since it's invoked only on pin change interrupt, so the axis grinds. All this to say we ultimately need checks when the block is loaded, or need to force endstop polling (which is what an earlier version of FTM used to do).

The timer based method doesn't impact stepper() runtime at all, and works with the endstop interrupts feature. I pushed it [here](https://github.com/S2AUlendo/MarlinCollaborative/commit/e589e859d8de7da0ba8da01af2c188536625fd26) since I'm working on top of some other changes/features to FTM, but the delta for this commit should be useful for your consideration.

I have tested it working on a few printers:
| Kinematics| Endstop switch type | Endstop monitor type |
| ---      | ---       | ---      |
| Cartesian | Sensorless XYZ | Polling |
| Cartesian | Sensorless XY, switch Z | Polling |
| Cartesian | Switch XY, BLTouch Z* | Pin Change Interrupt |
| CoreXY | Sensorless XYZ | Polling |

*Tested HS_MODE as well.

**Comment by narno2202** [comment]
I update my fork and will test the changes.
@ulendoalex , can you please test this PR with the 2 files provided. I just had a flag when` homeaxis() `is called, send `endstops.update() `in `stepper.cpp` only when the flag is true. After homing, endstops are managed by the temperature isr (polling) : on my printer it's OK with this PR and with the mix (previous post).
[stepper-motion.zip](https://github.com/MarlinFirmware/Marlin/files/14593369/stepper-motion.zip)

**Comment by narno2202** [comment]
@ulendoalex , I have tested your changes and everything seems to work fine. I still believe that there's no need to use `endstops.update()` after homing if temperature ISR polling is enough. Waiting for @thinkyhead  opinion.

**Comment by thinkyhead** [comment]
I just went ahead and cherry-picked e589e859d8de7da0ba8da01af2c188536625fd26 and adjusted the code. Do all the reversions and changes from that single commit apply? Let's retain the changes that make sense to get things stabilized and working and we can merge this for further optimization in additional PRs.

**Comment by narno2202** [comment]
I test it after adding the missing code, each axis move well in only in one direction, the opposite doesn't move at all. Previous code is working well for me as the code in #26848

**Comment by Mjankor** [comment]
Got this working on my sensorless CoreXY last night. Took a couple of "restore to defaults" to get everything behaving properly, and had to bump the buffer up ~2.5x to prevent misbehaviour on long, fast G0 moves.:
```
#define FTM_STEPPER_FS          75000
 #define FTM_STEPPERCMD_BUFF_SIZE 15000 //2.5x buffer size over default.
```


Still trying a few tweaks.

**Comment by thinkyhead** [comment]
> I test it after adding the missing code, each axis move well in only in one direction

@narno2202 — If you have a patch, please add it. I figured some of those reverted sections were inappropriate. We just need to decide which ones to keep and which to discard. Once we have that part sorted out, maybe this will be good to merge (?) for wider testing.

**Comment by Mjankor** [comment]
A couple of queries. 
The noise profile has changed, with a bit of a whine occuring on long, fast moves. Is this noise likely related to:
  #define FTM_FS                     1000         // (Hz) Frequency for trajectory generation. (Reciprocal of FTM_TS)

What are the ramifications to buffer size and printer behaviour if I change that number (and the corresponding FTM_TS)?

And what do I need to consider when changing these. 
  #define FTM_BW_SIZE               100         // Unified Window and Batch size with a ratio of 2
  #define FTM_STEPS_PER_LOOP           60         // Number of stepper commands to generate each loop()
  #define FTM_POINTS_PER_LOOP         100         // Number of trajectory points to generate each loop()
  
  There is also an obvious maximum buffer size that on my machine (Octopus Pro 429) comes in at around:
    #define FTM_STEPPER_FS          90000
    #define FTM_STEPPERCMD_BUFF_SIZE 18000 //3x buffer size over default.
    
When I try those values I start seeing stuttering on moves. Is that simply running out of memory or something else?

**Comment by narno2202** [comment]
Due to some sad familial problem, i am not very focused on Marlin these past days and for some days in the future. I will continue to check the repository.
@Mjankor  : you reach the limit of your motherboard with the last values.
In the new PR, buffer system mangement is a little bit different and seems to work well on ly test printer (not a core XY).

@thinkyhead , i don't have any patch for the described issue. I think that the new buffer management system is better than the current one. PR #26848 is a mix between last @ulendoalex code and the last working code of this PR. I'm still thinking that if Temperature ISR polling is enough to check endstops in case of disaster once the printer has homed (motions are in theory constrained within the bed size), so we can use `endstops.update()` only for homing and avoid unnecessary tests. You're the man with the answer :)

**Comment by thinkyhead** [comment]
I want to go ahead and merge our current work for further refinement. We still mark this "experimental," but it is reaching the point of being good enough for common usage. We'll continue to tune those configuration defaults as well and provide good defaults for our existing example machines. Thank you to all contributors for your attention and I hope that things are going better for everyone!

**Pull request merged**